server/src/main/java/io/crate/expression/predicate/NotPredicate.java

```diff
@@ -123,62 +110,64 @@ void add(Reference symbol) {
 
-        /**
-        * Three valued logic systems are defined in logic as systems in which there are 3 truth values: true,
-        * false and an indeterminate third value (in our case null is the third value).
-        * <p>
-        * This is a set of functions for which inputs evaluating to null needs to be explicitly included or
-        * excluded (in the case of 'not ...') in the boolean queries
-        */
-        private final Set<String> STRICT_3VL_FUNCTIONS =
-            Set.of(
-                AnyEqOperator.NAME,
-                AnyNeqOperator.NAME,
-                AndOperator.NAME,
-                AnyRangeOperator.Comparison.GT.opName(),
-                AnyRangeOperator.Comparison.GTE.opName(),
-                AnyRangeOperator.Comparison.LT.opName(),
-                AnyRangeOperator.Comparison.LTE.opName(),
-                LikeOperators.ANY_LIKE,
-                LikeOperators.ANY_NOT_LIKE,
-                CoalesceFunction.NAME,
-                CaseFunction.NAME,
-                ConcatFunction.NAME,
-                ConcatWsFunction.NAME,
-                ArrayCatFunction.NAME,
-                ArrayAppendFunction.NAME,
-                ArrayFunction.NAME,
-                ArrayUniqueFunction.NAME,
-                FormatFunction.NAME,
-                IfFunction.NAME,
-                IsNullPredicate.NAME,
-                MapFunction.NAME,
-                TryCastFunction.NAME,
-                SubscriptObjectFunction.NAME
-            );
+        private final Set<String> CAST_FUNCTIONS = Set.of(ImplicitCastFunction.NAME, ExplicitCastFunction.NAME);
 
         @Override
-        public Void visitReference(Reference symbol, SymbolToNotNullContext context) {
+        public Void visitReference(Reference symbol, NullabilityContext context) {
             context.add(symbol);
             return null;
         }
 
         @Override
-        public Void visitFunction(Function function, SymbolToNotNullContext context) {
+        public Void visitFunction(Function function, NullabilityContext context) {
             String functionName = function.name();
-            if (Ignore3vlFunction.NAME.equals(functionName)) {
-                return null;
-            }
-            if (!STRICT_3VL_FUNCTIONS.contains(functionName)) {
-                for (Symbol arg : function.arguments()) {
-                    arg.accept(this, context);
+            if (CAST_FUNCTIONS.contains(functionName)) {
+                // Cast functions should be ignored except for the case where the incoming
+                // datatype is an object. There we need to exclude null values to not match
+                // empty objects on the query
+                var a = function.arguments().get(0);
+                var b = function.arguments().get(1);
+                Reference ref = null;
+                if (a instanceof Reference aref && b instanceof Literal<?>) {
```

Comment:
This is a special case and required and not an optimization. If we cast to a `object` we need to require to add a field exist query to not match empty objects. Tests are breaking if this not set. I am not if this the right place for this logic.